lib/src/layer/polyline_layer.dart

```diff
@@ -39,66 +69,191 @@ class Polyline {
-    final map = MapCamera.of(context);
+    final camera = MapCamera.of(context);
 
     return MobileLayerTransformer(
       child: CustomPaint(
-        painter: PolylinePainter(
-          polylineCulling
-              ? polylines
-                  .where((p) => p.boundingBox.isOverlapping(map.visibleBounds))
-                  .toList()
-              : polylines,
-          map,
+        painter: _PolylinePainter<TapKeyType>(
+          polylines: polylines
+              .where((p) => p.boundingBox.isOverlapping(camera.visibleBounds))
+              .toList(),
+          camera: camera,
+          hitNotifier: hitNotifier,
+          onTapTolerance: hitTolerance,
         ),
-        size: Size(map.size.x, map.size.y),
+        size: Size(camera.size.x, camera.size.y),
         isComplex: true,
       ),
     );
   }
 }
 
-class PolylinePainter extends CustomPainter {
-  final List<Polyline> polylines;
-
-  final MapCamera map;
+class _PolylinePainter<TapKeyType extends Object> extends CustomPainter {
+  final List<Polyline<TapKeyType>> polylines;
+  final MapCamera camera;
   final LatLngBounds bounds;
+  final PolylineHitNotifier? hitNotifier;
+  final double onTapTolerance;
 
-  PolylinePainter(this.polylines, this.map) : bounds = map.visibleBounds;
+  _PolylinePainter({
+    required this.polylines,
+    required this.camera,
+    required this.hitNotifier,
+    required this.onTapTolerance,
+  }) : bounds = camera.visibleBounds;
 
-  int get hash => _hash ??= Object.hashAll(polylines);
+  final hits = List<Polyline<TapKeyType>>.empty(growable: true);
 
+  int get hash => _hash ??= Object.hashAll(polylines);
   int? _hash;
 
+  @override
+  bool? hitTest(Offset position) {
+    if (hitNotifier == null) return null;
+
+    hits.clear();
+
+    final origin =
+        camera.project(camera.center).toOffset() - camera.size.toOffset() / 2;
+
+    for (final p in polylines.reversed) {
+      // TODO: For efficiency we'd ideally filter by bounding box here. However
+      // we'd need to compute an extended bounding box that accounts account for
+      // the stroke width.
+      // if (!p.boundingBox.contains(touch)) {
+      //   continue;
+      // }
+
+      final offsets = getOffsets(camera, origin, p.points);
+      final strokeWidth = p.useStrokeWidthInMeter
+          ? _metersToStrokeWidth(
+              origin,
+              p.points.first,
+              offsets.first,
+              p.strokeWidth,
+            )
+          : p.strokeWidth;
+      final maxDistance =
+          (strokeWidth / 2 + p.borderStrokeWidth / 2) + onTapTolerance;
+
+      for (int i = 0; i < offsets.length - 1; i++) {
+        final o1 = offsets[i];
+        final o2 = offsets[i + 1];
+
+        final distance = math.sqrt(_distToSegmentSquared(
+          position.dx,
+          position.dy,
+          o1.dx,
+          o1.dy,
+          o2.dx,
+          o2.dy,
+        ));
+
+        if (distance < maxDistance) {
```

Comment:
No need to it now or ever but if we add back an occlusion parameter you "break outer" here and safe a few cycles.
That said, I do think that the above TODO would address any performance concerns more holistically, i.e. only look at polygons with an overlapping bbox

Reply:
Yeah, I'm aware of this trade-off, but wanted to minimize complicated configuration options.

The TODO seems more complicated and mathematically complex (converting stroke width to geographic space) than the work it would save. I'm hoping @mootw's upcoming PR #1704 will improve all this anyway, and will add culling within lines. Good review point for that tho, need to check whether it accounts for stroke width.

Reply:
> The TODO seems more complicated and mathematically complex (converting stroke width to geographic space) than the work it would save. 

We'll see but just FTR, I'd be surprised if that was more expensive. You can save a lot of segments but if #1704 provides the same benefit, even better.